crypto/evp/skeymgmt_meth.c

```diff
@@ -0,0 +1,236 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/crypto.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include "internal/core.h"
+#include "internal/provider.h"
+#include "internal/refcount.h"
+#include "crypto/evp.h"
+#include "evp_local.h"
+
+void *evp_skeymgmt_generate(const EVP_SKEYMGMT *skeymgmt, const OSSL_PARAM params[])
+{
+    void *provctx = ossl_provider_ctx(EVP_SKEYMGMT_get0_provider(skeymgmt));
+
+    return (skeymgmt->generate != NULL) ? skeymgmt->generate(provctx, params) : NULL;
+}
+
+void *evp_skeymgmt_import(const EVP_SKEYMGMT *skeymgmt, int selection, const OSSL_PARAM params[])
+{
+    void *provctx = ossl_provider_ctx(EVP_SKEYMGMT_get0_provider(skeymgmt));
+
+    /* This is mandatory, no need to check for its presence */
+    return skeymgmt->import(provctx, selection, params);
+}
+
+int evp_skeymgmt_export(const EVP_SKEYMGMT *skeymgmt, void *keydata,
+                        int selection, OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    /* This is mandatory, no need to check for its presence */
+    return skeymgmt->export(keydata, selection, param_cb, cbarg);
+}
+
+void evp_skeymgmt_freedata(const EVP_SKEYMGMT *skeymgmt, void *keydata)
+{
+    /* This is mandatory, no need to check for its presence */
+    skeymgmt->free(keydata);
+}
+
+static void *skeymgmt_new(void)
+{
+    EVP_SKEYMGMT *skeymgmt = NULL;
+
+    if ((skeymgmt = OPENSSL_zalloc(sizeof(*skeymgmt))) == NULL)
+        return NULL;
+    if (!CRYPTO_NEW_REF(&skeymgmt->refcnt, 1)) {
+        EVP_SKEYMGMT_free(skeymgmt);
+        return NULL;
+    }
+    return skeymgmt;
+}
+
+static void *skeymgmt_from_algorithm(int name_id,
+                                     const OSSL_ALGORITHM *algodef,
+                                     OSSL_PROVIDER *prov)
+{
+    const OSSL_DISPATCH *fns = algodef->implementation;
+    EVP_SKEYMGMT *skeymgmt = NULL;
+
+    if ((skeymgmt = skeymgmt_new()) == NULL)
+        return NULL;
+
+    skeymgmt->name_id = name_id;
+    if ((skeymgmt->type_name = ossl_algorithm_get1_first_name(algodef)) == NULL) {
+        EVP_SKEYMGMT_free(skeymgmt);
+        return NULL;
+    }
+    skeymgmt->description = algodef->algorithm_description;
+
+    for (; fns->function_id != 0; fns++) {
+        switch (fns->function_id) {
+        case OSSL_FUNC_SKEYMGMT_FREE:
+            if (skeymgmt->free == NULL)
+                skeymgmt->free = OSSL_FUNC_skeymgmt_free(fns);
+            break;
+        case OSSL_FUNC_SKEYMGMT_IMPORT:
+            if (skeymgmt->import == NULL)
+                skeymgmt->import = OSSL_FUNC_skeymgmt_import(fns);
+            break;
+        case OSSL_FUNC_SKEYMGMT_EXPORT:
+            if (skeymgmt->export == NULL)
+                skeymgmt->export = OSSL_FUNC_skeymgmt_export(fns);
+            break;
```

Comment:
If the import and export functions are mandatory as noted in your comments [here](https://github.com/openssl/openssl/pull/25908/files#r1951392106), then shouldn't a flag be set here, and checked below, to indicate that the OSSL_FUNC_SKEYMGMT_[EXPORT|IMPORT] methods have been set, similar to what we do in evp_asym_cipher_from_algorithm, with the ctxfncnt, encfcnt, etc, so that an error is returned if those methods aren't provided?  Doing so would catch an errant NULL pointer in a provider

Reply:
Look at line 113/114